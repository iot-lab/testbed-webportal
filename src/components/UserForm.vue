<template>
  <div>
    <div class="row">
      <div class="col-md">
        <div class="form-group" v-if="!hidden.includes('firstName')">
          <label class="form-control-label">First name</label>
          <input placeholder="First name" v-model="user.firstName" name="firstname"
            class="form-control" type="text"v-validate="'required'"
            :class="{'is-invalid': errors.has('firstname') }">
          <div class="invalid-feedback" v-show="errors.has('firstname')">
            {{ errors.first('firstname') }}
          </div>
        </div>
      </div>
      <div class="col-md">                    
        <div class="form-group" v-if="!hidden.includes('lastName')">
          <label class="form-control-label">Last name</label>
          <input placeholder="Last name" v-model="user.lastName" name="lastname"
            class="form-control" type="text" v-validate="'required'"
            :class="{'is-invalid': errors.has('lastname') }">
          <div class="invalid-feedback" v-show="errors.has('lastname')">
            {{ errors.first('lastname') }}
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col-md">                    
        <div class="form-group" v-if="!hidden.includes('email')">
          <label class="form-control-label">Email</label>
          <input v-model="user.email" name="email" class="form-control" type="email"
            placeholder="Academic or professional email" v-validate="'required|email|noWebMail'"
            :class="{'is-invalid': errors.has('email') }">
          <div class="invalid-feedback" v-show="errors.has('email')" v-html="errors.first('email')">
          </div>
        </div>
      </div>
      <div class="col-md">                    
      </div>
    </div>
    <div class="row">
      <div class="col-md">
        <div class="form-group" v-if="!hidden.includes('category')">
          <label class="form-control-label">User category</label>
          <multiselect v-model="user.category"
            placeholder="Category"
            :options="Object.keys(categories)"
            :allow-empty="false"
            :searchable="false"
            :show-labels="false"
            :custom-label="userLabel"
            :class="{'mymultiselect': true, 'invalid': dirty.category && !user.category}"
            @close="dirty.category = true">
          </multiselect>
          <div class="invalid-feedback" :style="{'display': (dirty.category && !user.category) ? 'block': 'none'}">
            The user category field is required.
          </div>
        </div>
      </div>
      <div class="col-md">
        <div class="form-group" v-if="!hidden.includes('organization')">
          <label class="form-control-label">Organization</label>
          <input placeholder="Organization" v-model="user.organization" name="organization"
            class="form-control" type="text" v-validate="'required'" :class="{'is-invalid': errors.has('organization') }">
          <div class="invalid-feedback" v-show="errors.has('organization')">
            {{ errors.first('organization') }}
          </div>
        </div>
      </div>
    </div>
    <div class="row">
      <div class="col-md">
        <div class="form-group" v-if="!hidden.includes('city')">
          <label class="form-control-label">City</label>
          <input placeholder="City" v-model="user.city" name="city"
            class="form-control" type="text" v-validate="'required'" :class="{'is-invalid': errors.has('city') }">
          <div class="invalid-feedback" v-show="errors.has('city')">
            {{ errors.first('city') }}
          </div>
        </div>
      </div>
      <div class="col-md">
        <div class="form-group" v-if="!hidden.includes('country')">
          <label class="form-control-label">Country</label>
          <multiselect v-model="user.country"
            placeholder="Country"
            :options="countries"
            :class="{'mymultiselect': true, 'invalid': dirty.country && !user.country}"
            @close="dirty.country = true">
          </multiselect>
          <div class="invalid-feedback" :style="{'display': (dirty.country && !user.country) ? 'block': 'none'}">
            The country field is required.
          </div>
        </div>
      </div>
    </div>
    <div class="form-group" v-if="!hidden.includes('motivations')">
      <label class="form-control-label">Motivations</label>
      <textarea v-model="user.motivations"
        name="motivations"
        class="form-control"
        rows="5"
        placeholder="Tell us about you"
        v-validate="'required'"
        @focus="toggleMotivations"
        @blur="toggleMotivations"
        :class="{'is-invalid': errors.has('motivations') }">
      </textarea>
      <div class="invalid-feedback" v-show="errors.has('motivations')">
        {{ errors.first('motivations') }}
      </div>
    </div>
    <div class="form-group" id="motivations-help" style="display: none" v-if="!hidden.includes('motivations-help')">
      <div class="alert alert-info form-control">
        <i class="fa fa-pencil"></i> <b>Tell us about your motivations</b>
        <ul style="margin: 0">
          <li>Research domain (Radio communication, networking protocol, distributed applications, &#x2026;)</li>
          <li>What kind of experiments do you want to run with IoT-LAB?</li>
          <li>Goal (Verify something existing in large scale, new development, &#x2026;)</li>
          <li>Experience with sensor networks (n00b, experiments with X platform, former IoT-LAB user, Guru, God)</li>
        </ul>
      </div>
    </div>
  </div>
</template>


<script>
import Multiselect from 'vue-multiselect'
import WebmailDomains from '@/assets/js/webmail-domains'
import countries from '@/assets/js/countries'
import UserCategories from '@/assets/js/categories'
import $ from 'jquery'
import { Validator } from 'vee-validate'

export default {
  name: 'UserForm',
  components: {Multiselect},

  props: {
    user: {
      type: Object,
      default: () => {},
    },
    hidden: {
      type: Object,
      default: () => [],
    },
  },

  data () {
    return {
      reCaptchaSitekey: '6Ld8cR4UAAAAAC-zBLP9m2bC35xyyYwTbvkBcx4q',
      countries: countries,
      categories: UserCategories,
      dirty: {
        category: false,
        country: false,
      },
    }
  },

  created () {
    Validator.extend('noWebMail', {
      getMessage: field => `Your email must be <b>academic</b> or <b>professional</b> in order to validate your account (<b>${this.user.email.split('@')[1]}</b> not allowed).`,
      validate: email => !WebmailDomains.includes(email.split('@')[1].toLowerCase()),
    })
  },

  methods: {
    async validate () {
      return this.$validator.validateAll().then((validated) => {
        if (!validated || !this.user.country || !this.user.category) {
          this.dirty.category = this.dirty.country = true
          console.log('User form is not valid.')
          return false
        }
        return true
      })
    },
    userLabel (cat) {
      return UserCategories[cat]
    },
    toggleMotivations () {
      $('#motivations-help').slideToggle()
    },
  },
}
</script>

<style scoped>
ul {
  padding-left: 20px;
}
</style>
