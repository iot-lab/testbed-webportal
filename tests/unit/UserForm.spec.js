import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import UserForm from '@/components/UserForm'
import { iotlab } from '@/rest'
jest.mock('@/rest')

// Cannot test using VeeValidate due to a bug, should try different lib versions or wait for a fix
// https://github.com/vuejs/vue-test-utils/issues/829
//
// workaround:
// localVue.use(VeeValidate)
const errors = {
  has: jest.fn(),
  first: jest.fn(),
}

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('UserForm.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(UserForm, {localVue, router, mocks: {errors}, propsData: {
      user: {"firstName":"Benoit","lastName":"Formet","email":"benoit.formet@inria.fr","organization":"Inria Saclay","city":"Paris","country":"France","login":"formet","motivations":"My goal is to make the iot-lab / testbed website shine","status":"active","created":"2016-12-16T09:49:09Z","category":"Academic","groups":["admin"],"sshkeys":["toto"]},
    }})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should match snapshot 2', () => {
    const wrapper = shallowMount(UserForm, {localVue, router, mocks: {errors}, propsData: {
      user: {"firstName":"Benoit","lastName":"Formet","email":"benoit.formet@inria.fr","organization":"Inria Saclay","city":"Paris","country":"France","login":"formet","motivations":"My goal is to make the iot-lab / testbed website shine","status":"active","created":"2016-12-16T09:49:09Z","category":"Academic","groups":["admin"],"sshkeys":["toto"]},
      hidden: ['firstName','lastName','email'],
      admin: true,
      mode: 'edit',
    }})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
