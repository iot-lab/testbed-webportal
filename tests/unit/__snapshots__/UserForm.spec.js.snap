// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm.vue should match snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          First name
        </label>
         
        <input
          class="form-control"
          name="firstname"
          placeholder="First name"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
     
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Last name
        </label>
         
        <input
          class="form-control"
          name="lastname"
          placeholder="Last name"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Email
        </label>
         
        <input
          class="form-control"
          name="email"
          placeholder="Academic or professional email"
          type="email"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        />
      </div>
    </div>
     
    <div
      class="col-md"
    />
  </div>
   
  <div
    class="row mb-5"
  >
    <div
      class="col form-group password"
    >
      <label
        class="form-control-label"
      >
        Password
      </label>
       
      <input
        class="form-control"
        name="password"
        placeholder="Password"
        type="password"
      />
       
      <div
        class="invalid-feedback"
        style="display: none;"
      >
        
        
      
      </div>
    </div>
     
    <div
      class="col form-group"
    >
      <label
        class="form-control-label"
      >
        Confirm password
      </label>
       
      <input
        class="form-control"
        data-vv-as="password"
        name="password_confirmation"
        placeholder="Password"
        type="password"
      />
       
      <div
        class="invalid-feedback"
        style="display: none;"
      >
        
        
      
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          User category
        </label>
         
        <multiselect-stub
          class="mymultiselect"
          custom-label="function () { [native code] }"
          deselectgrouplabel="Press enter to deselect group"
          deselectlabel="Press enter to remove"
          limit="99999"
          limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
          maxheight="300"
          name=""
          opendirection=""
          options="Student,Academic,Startup,Business"
          placeholder="Category"
          selectedlabel="Selected"
          selectgrouplabel="Press enter to select group"
          selectlabel="Press enter to select"
          shownoresults="true"
          tabindex="0"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          The user category field is required.
        
        </div>
      </div>
    </div>
     
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Organization
        </label>
         
        <input
          class="form-control"
          name="organization"
          placeholder="Organization"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          City
        </label>
         
        <input
          class="form-control"
          name="city"
          placeholder="City"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
     
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Country
        </label>
         
        <multiselect-stub
          class="mymultiselect"
          deselectgrouplabel="Press enter to deselect group"
          deselectlabel="Press enter to remove"
          limit="99999"
          limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
          maxheight="300"
          name=""
          opendirection=""
          options="Afghanistan,Åland Islands,Albania,Algeria,American Samoa,Andorra,Angola,Anguilla,Antarctica,Antigua and Barbuda,Argentina,Armenia,Aruba,Australia,Austria,Azerbaijan,Bahamas,Bahrain,Bangladesh,Barbados,Belarus,Belgium,Belize,Benin,Bermuda,Bhutan,Bolivia, Plurinational State of,Bonaire, Sint Eustatius and Saba,Bosnia and Herzegovina,Botswana,Bouvet Island,Brazil,British Indian Ocean Territory,Brunei Darussalam,Bulgaria,Burkina Faso,Burundi,Cambodia,Cameroon,Canada,Cape Verde,Cayman Islands,Central African Republic,Chad,Chile,China,Christmas Island,Cocos (Keeling) Islands,Colombia,Comoros,Congo,Congo, the Democratic Republic of the,Cook Islands,Costa Rica,Côte d'Ivoire,Croatia,Cuba,Curaçao,Cyprus,Czech Republic,Denmark,Djibouti,Dominica,Dominican Republic,Ecuador,Egypt,El Salvador,Equatorial Guinea,Eritrea,Estonia,Ethiopia,Falkland Islands (Malvinas),Faroe Islands,Fiji,Finland,France,French Guiana,French Polynesia,French Southern Territories,Gabon,Gambia,Georgia,Germany,Ghana,Gibraltar,Greece,Greenland,Grenada,Guadeloupe,Guam,Guatemala,Guernsey,Guinea,Guinea-Bissau,Guyana,Haiti,Heard Island and McDonald Islands,Holy See (Vatican City State),Honduras,Hong Kong,Hungary,Iceland,India,Indonesia,Iran, Islamic Republic of,Iraq,Ireland,Isle of Man,Israel,Italy,Jamaica,Japan,Jersey,Jordan,Kazakhstan,Kenya,Kiribati,Korea, Democratic People's Republic of,Korea, Republic of,Kuwait,Kyrgyzstan,Lao People's Democratic Republic,Latvia,Lebanon,Lesotho,Liberia,Libya,Liechtenstein,Lithuania,Luxembourg,Macao,Macedonia, the former Yugoslav Republic of,Madagascar,Malawi,Malaysia,Maldives,Mali,Malta,Marshall Islands,Martinique,Mauritania,Mauritius,Mayotte,Mexico,Micronesia, Federated States of,Moldova, Republic of,Monaco,Mongolia,Montenegro,Montserrat,Morocco,Mozambique,Myanmar,Namibia,Nauru,Nepal,Netherlands,New Caledonia,New Zealand,Nicaragua,Niger,Nigeria,Niue,Norfolk Island,Northern Mariana Islands,Norway,Oman,Pakistan,Palau,Palestinian Territory, Occupied,Panama,Papua New Guinea,Paraguay,Peru,Philippines,Pitcairn,Poland,Portugal,Puerto Rico,Qatar,Réunion,Romania,Russian Federation,Rwanda,Saint Barthélemy,Saint Helena, Ascension and Tristan da Cunha,Saint Kitts and Nevis,Saint Lucia,Saint Martin (French part),Saint Pierre and Miquelon,Saint Vincent and the Grenadines,Samoa,San Marino,Sao Tome and Principe,Saudi Arabia,Senegal,Serbia,Seychelles,Sierra Leone,Singapore,Sint Maarten (Dutch part),Slovakia,Slovenia,Solomon Islands,Somalia,South Africa,South Georgia and the South Sandwich Islands,South Sudan,Spain,Sri Lanka,Sudan,Suriname,Svalbard and Jan Mayen,Swaziland,Sweden,Switzerland,Syrian Arab Republic,Taiwan, Province of China,Tajikistan,Tanzania, United Republic of,Thailand,Timor-Leste,Togo,Tokelau,Tonga,Trinidad and Tobago,Tunisia,Turkey,Turkmenistan,Turks and Caicos Islands,Tuvalu,Uganda,Ukraine,United Arab Emirates,United Kingdom,United States,United States Minor Outlying Islands,Uruguay,Uzbekistan,Vanuatu,Venezuela, Bolivarian Republic of,Viet Nam,Virgin Islands, British,Virgin Islands, U.S.,Wallis and Futuna,Western Sahara,Yemen,Zambia,Zimbabwe"
          placeholder="Country"
          selectedlabel="Selected"
          selectgrouplabel="Press enter to select group"
          selectlabel="Press enter to select"
          showlabels="true"
          shownoresults="true"
          tabindex="0"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          The country field is required.
        
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <label
      class="form-control-label"
    >
      Motivations
    </label>
     
    <textarea
      class="form-control"
      name="motivations"
      placeholder="Tell us about you"
      rows="5"
    />
     
    <div
      class="invalid-feedback"
      style="display: none;"
    >
      
      
    
    </div>
  </div>
   
  <div
    class="form-group"
    id="motivations-help"
    style="display: none;"
  >
    <div
      class="alert alert-info form-control"
    >
      <i
        class="fa fa-pencil"
      />
       
      <b>
        Tell us about your motivations
      </b>
       
      <ul
        style="margin: 0px;"
      >
        <li>
          Research domain (Radio communication, networking protocol, distributed applications, …)
        </li>
         
        <li>
          What kind of experiments do you want to run with IoT-LAB?
        </li>
         
        <li>
          Goal (Verify something existing in large scale, new development, …)
        </li>
         
        <li>
          Experience with sensor networks (n00b, experiments with X platform, former IoT-LAB user, Guru, God)
        </li>
      </ul>
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <a
      aria-controls="collapseSSH"
      aria-expanded="false"
      data-toggle="collapse"
      href=".collapse-ssh"
      role="button"
    >
      <label
        class="form-control-label"
      >
        <i
          class="fa fa-caret-right"
        />
         add SSH key 
        <span
          class="text-muted"
        >
          (optional, can be done later)
        </span>
      </label>
    </a>
     
    <sshkeys-stub
      class="collapse collapse-ssh"
      keys="toto"
      rows="4"
    />
     
    <div
      class="invalid-feedback"
      style="display: none;"
    >
      
      
    
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <label
      class="form-control-label"
    >
      User groups
    </label>
     
    <multiselect-stub
      allow-empty="true"
      class="mymultiselect"
      deselectgrouplabel="Press enter to deselect group"
      deselectlabel="Press enter to remove"
      hide-selected="true"
      limit="99999"
      limittext="function (t){return\\"and \\".concat(t,\\" more\\")}"
      maxheight="300"
      multiple="true"
      name=""
      opendirection=""
      options=""
      placeholder="Groups"
      selectedlabel="Selected"
      selectgrouplabel="Press enter to select group"
      selectlabel="Press enter to select"
      shownoresults="true"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`UserForm.vue should match snapshot 2 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <!---->
    </div>
     
    <div
      class="col-md"
    >
      <!---->
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <!---->
    </div>
     
    <div
      class="col-md"
    />
  </div>
   
  <div
    class="row mb-5"
  >
    <div
      class="col form-group password"
    >
      <label
        class="form-control-label"
      >
        Password
      </label>
       
      <input
        class="form-control"
        name="password"
        placeholder="Password"
        type="password"
      />
       
      <div
        class="invalid-feedback"
        style="display: none;"
      >
        
        
      
      </div>
    </div>
     
    <div
      class="col form-group"
    >
      <label
        class="form-control-label"
      >
        Confirm password
      </label>
       
      <input
        class="form-control"
        data-vv-as="password"
        name="password_confirmation"
        placeholder="Password"
        type="password"
      />
       
      <div
        class="invalid-feedback"
        style="display: none;"
      >
        
        
      
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          User category
        </label>
         
        <multiselect-stub
          class="mymultiselect"
          custom-label="function () { [native code] }"
          deselectgrouplabel="Press enter to deselect group"
          deselectlabel="Press enter to remove"
          limit="99999"
          limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
          maxheight="300"
          name=""
          opendirection=""
          options="Student,Academic,Startup,Business"
          placeholder="Category"
          selectedlabel="Selected"
          selectgrouplabel="Press enter to select group"
          selectlabel="Press enter to select"
          shownoresults="true"
          tabindex="0"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          The user category field is required.
        
        </div>
      </div>
    </div>
     
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Organization
        </label>
         
        <input
          class="form-control"
          name="organization"
          placeholder="Organization"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          City
        </label>
         
        <input
          class="form-control"
          name="city"
          placeholder="City"
          type="text"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          
        
        </div>
      </div>
    </div>
     
    <div
      class="col-md"
    >
      <div
        class="form-group"
      >
        <label
          class="form-control-label"
        >
          Country
        </label>
         
        <multiselect-stub
          class="mymultiselect"
          deselectgrouplabel="Press enter to deselect group"
          deselectlabel="Press enter to remove"
          limit="99999"
          limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
          maxheight="300"
          name=""
          opendirection=""
          options="Afghanistan,Åland Islands,Albania,Algeria,American Samoa,Andorra,Angola,Anguilla,Antarctica,Antigua and Barbuda,Argentina,Armenia,Aruba,Australia,Austria,Azerbaijan,Bahamas,Bahrain,Bangladesh,Barbados,Belarus,Belgium,Belize,Benin,Bermuda,Bhutan,Bolivia, Plurinational State of,Bonaire, Sint Eustatius and Saba,Bosnia and Herzegovina,Botswana,Bouvet Island,Brazil,British Indian Ocean Territory,Brunei Darussalam,Bulgaria,Burkina Faso,Burundi,Cambodia,Cameroon,Canada,Cape Verde,Cayman Islands,Central African Republic,Chad,Chile,China,Christmas Island,Cocos (Keeling) Islands,Colombia,Comoros,Congo,Congo, the Democratic Republic of the,Cook Islands,Costa Rica,Côte d'Ivoire,Croatia,Cuba,Curaçao,Cyprus,Czech Republic,Denmark,Djibouti,Dominica,Dominican Republic,Ecuador,Egypt,El Salvador,Equatorial Guinea,Eritrea,Estonia,Ethiopia,Falkland Islands (Malvinas),Faroe Islands,Fiji,Finland,France,French Guiana,French Polynesia,French Southern Territories,Gabon,Gambia,Georgia,Germany,Ghana,Gibraltar,Greece,Greenland,Grenada,Guadeloupe,Guam,Guatemala,Guernsey,Guinea,Guinea-Bissau,Guyana,Haiti,Heard Island and McDonald Islands,Holy See (Vatican City State),Honduras,Hong Kong,Hungary,Iceland,India,Indonesia,Iran, Islamic Republic of,Iraq,Ireland,Isle of Man,Israel,Italy,Jamaica,Japan,Jersey,Jordan,Kazakhstan,Kenya,Kiribati,Korea, Democratic People's Republic of,Korea, Republic of,Kuwait,Kyrgyzstan,Lao People's Democratic Republic,Latvia,Lebanon,Lesotho,Liberia,Libya,Liechtenstein,Lithuania,Luxembourg,Macao,Macedonia, the former Yugoslav Republic of,Madagascar,Malawi,Malaysia,Maldives,Mali,Malta,Marshall Islands,Martinique,Mauritania,Mauritius,Mayotte,Mexico,Micronesia, Federated States of,Moldova, Republic of,Monaco,Mongolia,Montenegro,Montserrat,Morocco,Mozambique,Myanmar,Namibia,Nauru,Nepal,Netherlands,New Caledonia,New Zealand,Nicaragua,Niger,Nigeria,Niue,Norfolk Island,Northern Mariana Islands,Norway,Oman,Pakistan,Palau,Palestinian Territory, Occupied,Panama,Papua New Guinea,Paraguay,Peru,Philippines,Pitcairn,Poland,Portugal,Puerto Rico,Qatar,Réunion,Romania,Russian Federation,Rwanda,Saint Barthélemy,Saint Helena, Ascension and Tristan da Cunha,Saint Kitts and Nevis,Saint Lucia,Saint Martin (French part),Saint Pierre and Miquelon,Saint Vincent and the Grenadines,Samoa,San Marino,Sao Tome and Principe,Saudi Arabia,Senegal,Serbia,Seychelles,Sierra Leone,Singapore,Sint Maarten (Dutch part),Slovakia,Slovenia,Solomon Islands,Somalia,South Africa,South Georgia and the South Sandwich Islands,South Sudan,Spain,Sri Lanka,Sudan,Suriname,Svalbard and Jan Mayen,Swaziland,Sweden,Switzerland,Syrian Arab Republic,Taiwan, Province of China,Tajikistan,Tanzania, United Republic of,Thailand,Timor-Leste,Togo,Tokelau,Tonga,Trinidad and Tobago,Tunisia,Turkey,Turkmenistan,Turks and Caicos Islands,Tuvalu,Uganda,Ukraine,United Arab Emirates,United Kingdom,United States,United States Minor Outlying Islands,Uruguay,Uzbekistan,Vanuatu,Venezuela, Bolivarian Republic of,Viet Nam,Virgin Islands, British,Virgin Islands, U.S.,Wallis and Futuna,Western Sahara,Yemen,Zambia,Zimbabwe"
          placeholder="Country"
          selectedlabel="Selected"
          selectgrouplabel="Press enter to select group"
          selectlabel="Press enter to select"
          showlabels="true"
          shownoresults="true"
          tabindex="0"
        />
         
        <div
          class="invalid-feedback"
          style="display: none;"
        >
          
          The country field is required.
        
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <label
      class="form-control-label"
    >
      Motivations
    </label>
     
    <textarea
      class="form-control"
      name="motivations"
      placeholder="Tell us about you"
      rows="5"
    />
     
    <div
      class="invalid-feedback"
      style="display: none;"
    >
      
      
    
    </div>
  </div>
   
  <div
    class="form-group"
    id="motivations-help"
    style="display: none;"
  >
    <div
      class="alert alert-info form-control"
    >
      <i
        class="fa fa-pencil"
      />
       
      <b>
        Tell us about your motivations
      </b>
       
      <ul
        style="margin: 0px;"
      >
        <li>
          Research domain (Radio communication, networking protocol, distributed applications, …)
        </li>
         
        <li>
          What kind of experiments do you want to run with IoT-LAB?
        </li>
         
        <li>
          Goal (Verify something existing in large scale, new development, …)
        </li>
         
        <li>
          Experience with sensor networks (n00b, experiments with X platform, former IoT-LAB user, Guru, God)
        </li>
      </ul>
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <a
      aria-controls="collapseSSH"
      aria-expanded="false"
      data-toggle="collapse"
      href=".collapse-ssh"
      role="button"
    >
      <label
        class="form-control-label"
      >
        <i
          class="fa fa-caret-right"
        />
         SSH keys
      
      </label>
    </a>
     
    <sshkeys-stub
      class="collapse collapse-ssh"
      keys="toto"
      rows="4"
    />
     
    <div
      class="invalid-feedback"
      style="display: none;"
    >
      
      
    
    </div>
  </div>
   
  <div
    class="form-group"
  >
    <label
      class="form-control-label"
    >
      User groups
    </label>
     
    <multiselect-stub
      allow-empty="true"
      class="mymultiselect"
      deselectgrouplabel="Press enter to deselect group"
      deselectlabel="Press enter to remove"
      hide-selected="true"
      limit="99999"
      limittext="function (t){return\\"and \\".concat(t,\\" more\\")}"
      maxheight="300"
      multiple="true"
      name=""
      opendirection=""
      options=""
      placeholder="Groups"
      selectedlabel="Selected"
      selectgrouplabel="Press enter to select group"
      selectlabel="Press enter to select"
      shownoresults="true"
      tabindex="0"
    />
  </div>
</div>
`;
